Add classes to BMSBK

dyceron opened the pull request:
Now has classes with properties to better handle modifying `block_groups`

duncathan commented on tests/formats/test_bmsbk.py:
```diff
@@ -25,15 +25,20 @@ def surface_bmsbk(samus_returns_tree) -> Bmsbk:
 
 
 def test_get_block(surface_bmsbk: Bmsbk):
-    block = surface_bmsbk.get_block(0, 1)
```
your test is using a different block id now, is that intended?

duncathan commented on tests/formats/test_bmsbk.py:
```diff
@@ -25,15 +25,20 @@ def surface_bmsbk(samus_returns_tree) -> Bmsbk:
+    orignal_type = block_group.block_type
+    block_group.block_type == BlockType.BOMB
+    assert orignal_type != BlockType.BOMB
 
```
do this assert *before* changing the type, and then do another assert afterwards to verify that the new value is what you set it to

duncathan commented on tests/formats/test_bmsbk.py:
```diff
@@ -25,15 +25,20 @@ def surface_bmsbk(samus_returns_tree) -> Bmsbk:
+    block = surface_bmsbk.get_block_group(0).get_block(0)
+    original_time = block.respawn_time
+    block.respawn_time == 5.0
+    assert original_time != 5.0
```
same thing as the other test

duncathan commented on tests/formats/test_bmsbk.py:
```diff
@@ -25,15 +25,20 @@ def surface_bmsbk(samus_returns_tree) -> Bmsbk:
-    assert surface_bmsbk.get_block_group(1).types[0].block_type == BlockType.BOMB
+    block_group = surface_bmsbk.get_block_group(1)
+    orignal_type = block_group.block_type
+    block_group.block_type == BlockType.BOMB
```
I assume you mean `=` here and in the other test

duncathan reviewed (changes requested):
please add tests for the other properties to get full code coverage

duncathan commented on src/mercury_engine_data_structures/formats/bmsbk.py:
```diff
@@ -55,15 +55,12 @@ def _rebuild_types(ctx: Container) -> int:
 
 
-class BlockType(Enum):
+class BlockType(StrEnum):
```
StrEnum breaks python 3.10 - do `class BlockType(str, Enum)` instead

duncathan commented on tests/formats/test_bmsbk.py:
```diff
@@ -25,15 +25,40 @@ def surface_bmsbk(samus_returns_tree) -> Bmsbk:
+    original_type = block_group.block_type
+    assert original_type == BlockType.POWER_BOMB
+    block_group.block_type = BlockType.BOMB
+    assert original_type != BlockType.BOMB
```
```py
block_group = surface_bmsbk.get_block_group(0)
assert block_group.block_type is BlockType.POWER_BOMB
block_group.block_type = BlockType.BOMB
assert block_group.block_type is BlockType.BOMB
```

duncathan commented on tests/formats/test_bmsbk.py:
```diff
@@ -25,15 +25,40 @@ def surface_bmsbk(samus_returns_tree) -> Bmsbk:
+    original_time = block.respawn_time
+    assert original_time == 0.0
+    block.respawn_time = 5.0
+    assert original_time != block.respawn_time
```
```py
block = surface_bmsbk.get_block_group(0).get_block(0)
assert block.respawn_time == 0.0
block.respawn_time = 5.0
assert block.respawn_time == 5.0
```

duncathan commented on tests/formats/test_bmsbk.py:
```diff
@@ -25,15 +25,40 @@ def surface_bmsbk(samus_returns_tree) -> Bmsbk:
+    assert block.model_name != "sg_casca80"
+
+    block.vignette_name = "sg_real_vignette"
+    assert block.vignette_name != ""
```
```py
block = surface_bmsbk.get_block_group(0).get_block(0)

assert block.model_name == "sd_casca80"
block.model_name = "sg_real_model"
assert block.model_name == "sg_real_model"

assert block.vignette_name == ""
block.vignette_name = "sg_real_vignette"
assert block.vignette_name == "sg_real_vignette"
```

duncathan commented on tests/formats/test_bmsbk.py:
```diff
@@ -25,15 +25,40 @@ def surface_bmsbk(samus_returns_tree) -> Bmsbk:
+    original_position = block.position
+    assert original_position == [-23100.0, 10700.0, 0.0]
+    block.position = [100.0, 200.0, 0.0]
+    assert original_position != block.position
```
```py
block = surface_bmsbk.get_block_group(0).get_block(0)

assert block.position == [-23100.0, 10700.0, 0.0]
block.position = Vec3(100.0, 200.0, 0.0)
assert block.position == [100.0, 200.0, 0.0]
```

dyceron commented:
- `block_type` now uses and enum for decoding
- Added functions to add blocks, block groups, and collision_camera_groups
- Added function to remove all blocks from a collision_camera as well as the collision_camera itself

duncathan commented on src/mercury_engine_data_structures/formats/bmsbk.py:
```diff
@@ -41,37 +43,108 @@ def _rebuild_types(ctx: Container) -> int:
+    def get_block(self, block_idx: int) -> BlockData:
+        return BlockData(self._raw.types[0].blocks[block_idx])
+
+    def add_block(
+        self, position: Vec3, respawn_time: float, model_name: str, vignette_name: str, unk2: int = 0
+    ) -> Container:
+        new_block = Container(
+            {
+                "position": position,
+                "unk2": unk2,
+                "unk3": 0,
+                "respawn_time": respawn_time,
+                "model_name": model_name,
+                "vignette_name": vignette_name,
+            }
+        )
```
this should probably be a classmethod on `BlockData`:
```py
class BlockData:
    @classmethod
    def create(cls, position: Vec3, respawn_time: float, model_name: str, vignette_name: str, unk2: int = 0) -> typing.Self:
        return cls(Container({
            "position": position,
            "unk2": unk2,
            "unk3": 0,
            "respawn_time": respawn_time,
            "model_name": model_name,
            "vignette_name": vignette_name,
        }))
```

dyceron replied:
Ah I've never used classmethod before. Seems cleaner.

duncathan commented on src/mercury_engine_data_structures/formats/bmsbk.py:
```diff
@@ -41,37 +43,108 @@ def _rebuild_types(ctx: Container) -> int:
+            }
+        )
+
+        self._raw.types[0].blocks.append(new_block)
```
and then this function can just be this:
```py
def add_block(self, new_block: BlockData) -> None:
    self._raw.types[0].blocks.append(new_block)
```
